File: src/main/java/cz/nitramek/messaging/network/UDPSender.java
package cz.nitramek.messaging.network;


import lombok.ToString;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

@ToString
@Slf4j
public class UDPSender extends UDPWorker {

    private BlockingQueue<Packet> packets;

    public UDPSender() {
        packets = new ArrayBlockingQueue<>(10);
    }

    @Override
    public void work(DatagramChannel channel) throws Exception {
        Packet packet = packets.poll(100, TimeUnit.MILLISECONDS);
        if (packet != null) {
            log.debug("Sending message to {} of size {}", packet.getRecipient(), packet.getBuffer().length);
            ByteBuffer buffer = ByteBuffer.wrap(packet.getBuffer());
            channel.send(buffer, packet.getRecipient());
        }
    }

    public void sendPacket(Packet packet) {
        packets.add(packet);
    }

    @Override
    public void shutdown() throws Exception {
        super.shutdown();
    }
}

File: src/main/java/cz/nitramek/messaging/network/UDPSender.kt
package cz.nitramek.messaging.network


import cz.nitramek.agent.SENDER_THREAD_COUNT
import org.slf4j.LoggerFactory
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel
import java.util.concurrent.Executors
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.TimeUnit

class UDPSender(threadCount: Int = SENDER_THREAD_COUNT) {

    private val log = LoggerFactory.getLogger(this::class.java)!!
    private val messagesLog = LoggerFactory.getLogger("receivedMessages")!!

    private var channel: DatagramChannel? = null
    private val pool = Executors.newScheduledThreadPool(threadCount)

    fun start() {
        channel = DatagramChannel.open()
    }

    private fun sendMessage(str: String, address: InetSocketAddress) {
        log.debug("Sending message to {} of {}", address, str)
        val buffer = ByteBuffer.wrap(str.toByteArray(Charsets.UTF_8))
        channel?.send(buffer, address)


    }


    fun sendPacket(str: String, address: InetSocketAddress) {
        pool.schedule({
            sendImmediadly(str, address)
        }, ThreadLocalRandom.current().nextLong(10), TimeUnit.MILLISECONDS)
    }

    fun sendImmediadly(str: String, address: InetSocketAddress) {
        sendMessage(str, address)
        messagesLog.trace("Sending {} to", str, address)
    }

    fun shutdown() {
        pool.awaitTermination(100, TimeUnit.MILLISECONDS)
        pool.shutdownNow()
        channel?.close()

    }

}
